lib/fswatcher/fswatcher.go

```
@@ -0,0 +1,558 @@
+	notifyDelay time.Duration
+	// time after which an active event is passed to syncthing-core
+	notifyTimeout         time.Duration
+	notifyTimer           *time.Timer
+	notifyTimerNeedsReset bool
+	notifyTimerResetChan  chan time.Duration
+	notifyChan            chan []string
+	// All detected and to be scanned events are stored in a tree like
+	// structure mimicking folders to keep count of events per directory.
+	rootEventDir     *eventDir
+	backendEventChan chan notify.EventInfo
+	inProgress       map[string]struct{}
+	cfg              *config.Wrapper
+	configUpdate     chan config.FolderConfiguration
+	stop             chan struct{}
+}
+
+type Service interface {
+	Serve()
+	Stop()
+	C() <-chan []string
+	UpdateIgnores(ignores *ignore.Matcher)
+	VerifyConfiguration(from, to config.Configuration) error
+	CommitConfiguration(from, to config.Configuration) bool
+	String() string
+}
+
+func New(id string, cfg *config.Wrapper, ignores *ignore.Matcher) (Service, error) {
+	fsWatcher := &watcher{
+		folderID:              id,
+		folderIgnores:         ignores,
+		folderIgnoresUpdate:   make(chan *ignore.Matcher),
+		notifyTimerNeedsReset: false,
+		notifyTimerResetChan:  make(chan time.Duration),
+		notifyChan:            make(chan []string),
+		rootEventDir:          newEventDir(".", nil),
+		backendEventChan:      make(chan notify.EventInfo, maxFiles),
+		inProgress:            make(map[string]struct{}),
+		cfg:                   cfg,
+		configUpdate:          make(chan config.FolderConfiguration),
+		stop:                  make(chan struct{}),
+	}
+	folderCfg, ok := cfg.Folder(id)
+	if !ok {
+		panic(fmt.Sprintf("bug: Folder %s does not exist", id))
+	}
+	fsWatcher.updateConfig(folderCfg)
+
+	if err := fsWatcher.setupBackend(); err != nil {
+		return nil, err
+	}
+
+	return fsWatcher, nil
+}
+
+func (w *watcher) setupBackend() error {
+	absShouldIgnore := func(absPath string) bool {
+		if !isSubpath(absPath, w.folderPath) {
+			return true
```

I think it shouldn't happen, yes, but it is not in our direct control, as the arguments come from the backend. However I guess it is a good idea, otherwise a problem here could silently drop all changes.